### PR TITLE
feat(cc): implement Thermostat Fan Mode CC

### PR DESCRIPTION
**Pull request by Ikcelaks**
I've created a Draft implementation of the Thermostat Fan Mode CC and an incomplete test suite.

The implementation should be complete, but I haven't yet tested it on actual hardware.  I'm submitting this PR as a draft to get early feedback on whether you're interested in me developing it further.

(I also intend to add the Thermostat Fan State CC and expand the tests)

### CODE REVIEW

File: packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+
+		this.persistValues();
+	}
+
+	private _mode: ThermostatFanMode;
+	@ccValue()
+	@ccValueMetadata({
+		...ValueMetadata.UInt8,
+		min: 0,
+		max: 15,
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts]: This should be replaced with `states: enumValuesToMetadataStates(ThermostatFanMode)`, by default don't limit it to any modes. You're doing that in the `SupportedReport`.

File: packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+		label: "Thermostat fan mode",
+	})
+	public get mode(): ThermostatFanMode {
+		return this._mode;
+	}
+
+	private _off: boolean | undefined;
+	@ccValue({ minVersion: 2 })
+	public get off(): boolean | undefined {
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts:375]: missing metadata

File: packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+			this.mode = options.mode;
+		}
+	}
+
+	public mode: ThermostatFanMode;
+
+	public serialize(): Buffer {
+		this.payload = Buffer.from([
+			// 0 R R R M M M M
+			this.mode & 0b1111,
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts]: you're missing the `off` flag

File: packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+			endpoint: this.endpointIndex,
+			property: "mode",
+		};
+		// Only update the dynamic part
+		this.getValueDB().setMetadata(valueId, ({
+			states: enumValuesToMetadataStates(
+				ThermostatFanMode,
+				this._supportedModes,
+			),
+		} as unknown) as ValueMetadata);
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts]: It is not ideal, but you can get around the double cast by doing this:
```ts
		this.getValueDB().setMetadata(valueId, {
			...ValueMetadata.UInt8, // <== HERE!
			states: enumValuesToMetadataStates(
				ThermostatFanMode,
				this._supportedModes,
			),
		});
```

File: packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+@CCCommand(ThermostatFanModeCommand.SupportedReport)
+export class ThermostatFanModeCCSupportedReport extends ThermostatFanModeCC {
+	public constructor(
+		driver: Driver,
+		options: CommandClassDeserializationOptions,
+	) {
+		super(driver, options);
+		this._supportedModes = parseBitMask(
+			this.payload,
+			ThermostatFanMode["Auto high"],
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts]:
```suggestion
			ThermostatFanMode["Auto low"],
```

File: packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+			// Verify the current value after a delay
+			this.schedulePoll({ property });
+		}
+	};
+
+	protected [POLL_VALUE]: PollValueImplementation = async ({
+		property,
+	}): Promise<unknown> => {
+		switch (property) {
+			case "mode":
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts:151]:
```suggestion
			case "mode":
			case "off":
```

File: packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+		}
+		if (typeof value !== "number") {
+			throwWrongValueType(this.ccId, property, "number", typeof value);
+		}
+		await this.set(value);
+
+		if (this.isSinglecast()) {
+			// Verify the current value after a delay
+			this.schedulePoll({ property });
+		}
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts]: This should also allow setting the `"off"` property. In that case, I'd load the current mode from the value DB and use it together with `off`.

File: packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+			message: "querying current thermostat fan mode...",
+			direction: "outbound",
+		});
+		const currentStatus = await api.get();
+		if (currentStatus) {
+			this.driver.controllerLog.logNode(node.id, {
+				endpoint: this.endpointIndex,
+				message:
+					"received current thermostat fan mode: " +
+					getEnumMemberName(ThermostatFanMode, currentStatus.mode),
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts]: I'd log `off` too.

**Review by AlCalzone — CHANGES_REQUESTED** [review]
Looks pretty good - a few minor things need to be addressed

File: packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts (hunk 7 above)
Comment (Ikcelaks) [reply packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts]: I just reread the specs on the Set command, and you're of course correct. I do think they really could have written that paragraph better.

File: packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts (hunk 1 above)
Comment (Ikcelaks) [reply packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts]: I essentially copied the metadata info from the ThermostatModeCC.ts file (which has min 0 and max 31), so I thought it was a guard on the size of the bitmask.  Might want to look at that CC too.

File: packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts (hunk 4 above)
Comment (Ikcelaks) [reply packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts]: I also copied this bit from ThermostatModeCC.ts. This looks much saner.

File: packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts (hunk 7 above)
Comment (AlCalzone) [reply packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts]: Sounds familiar 😅

**Comment by Ikcelaks** [comment]
I'm very unsure that this is handling the set and report correctly with respect to the Off bit.  In particular, I think that setting the Off bit to 0 without specifying the Mode is nonsensical.

My device doesn't have an "off" mode, so I will not be able to test if this is correct with any hardware.

File: packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts (hunk 1 above)
Comment (AlCalzone) [reply packages/zwave-js/src/lib/commandclass/ThermostatFanModeCC.ts]: Good point!

**Comment by AlCalzone** [comment]
>  In particular, I think that setting the Off bit to 0 without specifying the Mode is nonsensical.

I'm not sure about this. Doesn't it make sense to turn the device back on without changing the mode?

**Review by AlCalzone — DISMISSED** [review]

**Review by AlCalzone — APPROVED** [review]

**Pull request merged**